### ISSUE TITLE
[rocky9] scylla generate phantom coredump on calls to `scylla --version`

### ISSUE THREAD

**Issue by fruch** [scylladb/scylladb#13945]
## Issue description

- [ ] This issue is a regression.
- [x] It is unknown if this issue is a regression.

offline install on rocky9, scylla seems to be generating empty coredumps
```
May 18 11:25:07 artifacts-rocky9-jenkins-db-node-28b4b2b0-0-1 systemd[1]: Started Process Core Dump (PID 67237/UID 0).
May 18 11:25:07 artifacts-rocky9-jenkins-db-node-28b4b2b0-0-1 systemd-coredump[67238]: Resource limits disable core dumping for process 67223 (scylla).
May 18 11:25:07 artifacts-rocky9-jenkins-db-node-28b4b2b0-0-1 systemd-coredump[67238]: [🡕] Process 67223 (scylla) of user 1000 dumped core.
May 18 11:25:07 artifacts-rocky9-jenkins-db-node-28b4b2b0-0-1 systemd[1]: systemd-coredump@0-67237-0.service: Deactivated successfully.
May 18 11:25:10 artifacts-rocky9-jenkins-db-node-28b4b2b0-0-1 systemd[1]: Started Process Core Dump (PID 67342/UID 0).
May 18 11:25:10 artifacts-rocky9-jenkins-db-node-28b4b2b0-0-1 systemd-coredump[67343]: Resource limits disable core dumping for process 67328 (scylla).
May 18 11:25:10 artifacts-rocky9-jenkins-db-node-28b4b2b0-0-1 systemd-coredump[67343]: [🡕] Process 67328 (scylla) of user 1000 dumped core.
May 18 11:25:10 artifacts-rocky9-jenkins-db-node-28b4b2b0-0-1 systemd[1]: systemd-coredump@1-67342-0.service: Deactivated successfully.
```  

those core reports doesn't actually has any corefile, and scylla seems to be running o.k. after it
```
[scylla-test@artifacts-rocky9-jenkins-db-node-28b4b2b0-0-1 ~]$ coredumpctl
TIME                          PID  UID  GID SIG     COREFILE EXE                                       SIZE
Thu 2023-05-18 11:25:07 UTC 67223 1000 1001 SIGABRT none     /home/scylla-test/scylladb/libexec/scylla  n/a
Thu 2023-05-18 11:25:10 UTC 67328 1000 1001 SIGABRT none     /home/scylla-test/scylladb/libexec/scylla  n/a
```
those phantom coredump seems to be matching the calls to `scylla --version`

```
< t:2023-05-18 11:25:10,646 f:remote_base.py  l:520  c:RemoteLibSSH2CmdRunner p:DEBUG > Running command "/home/scylla-test/scylladb/bin/scylla --version"...
< t:2023-05-18 11:25:11,148 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > terminate called after throwing an instance of 'std::runtime_error'
< t:2023-05-18 11:25:11,649 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG >   what():  Could not setup Async I/O: Resource temporarily unavailable. The most common cause is not enough request capacity in /proc/sys/fs/aio-max-nr. Try increasing that number or reducing the amount of logical CPUs available for your application
< t:2023-05-18 11:25:11,649 f:base.py         l:140  c:RemoteLibSSH2CmdRunner p:DEBUG > STDERR: terminate called after throwing an instance of 'std::runtime_error'
< t:2023-05-18 11:25:11,649 f:base.py         l:140  c:RemoteLibSSH2CmdRunner p:DEBUG >   what():  Could not setup Async I/O: Resource temporarily unavailable. The most common cause is not enough request capacity in /proc/sys/fs/aio-max-nr. Try increasing that number or reducing the amount of logical CPUs available for your application
< t:2023-05-18 11:25:11,649 f:base.py         l:142  c:RemoteLibSSH2CmdRunner p:DEBUG > Command "/home/scylla-test/scylladb/bin/scylla --version" finished with status 0
```

It's quite weird that scylla is running just fine, and `scylla ---version` compiling about `/proc/sys/fs/aio-max-nr`

```
[scylla-test@artifacts-rocky9-jenkins-db-node-28b4b2b0-0-1 ~]$ cat /proc/sys/fs/aio-max-nr
65536
```

## Impact

user might be confused to think scylla is crashing even that doesn't seem to be the case 

## How frequently does it reproduce?

seems like it's reproducible and show the exact same behavior 3 times in a row

note this is a new test case (i.e. support for rocky), and we don't have lots of history since it was failing related to JMX issues until now

## Installation details

Cluster size: 1 nodes (n1-standard-2)

Scylla Nodes used in this run:
- artifacts-rocky9-jenkins-db-node-28b4b2b0-0-1 (34.23.199.235 | 10.142.0.12) (shards: 2)

OS / Image: `https://www.googleapis.com/compute/v1/projects/rocky-linux-cloud/global/images/family/rocky-linux-9` (gce: us-east1)

Test: `artifacts-rocky9-nonroot-test`
Test id: `28b4b2b0-149c-49d2-bacf-74c80e9a51d7`
Test name: `scylla-master/artifacts-offline-install/artifacts-rocky9-nonroot-test`
Test config file(s):

- [rocky9.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/67ca0c1206af4fb2ca286369e1a791a2fba35c1a/test-cases/artifacts/rocky9.yaml)

<details>
<summary>
Logs and commands
</summary>

- Restore Monitor Stack command: `$ hydra investigate show-monitor 28b4b2b0-149c-49d2-bacf-74c80e9a51d7`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=28b4b2b0-149c-49d2-bacf-74c80e9a51d7)
- Show all stored logs command: `$ hydra investigate show-logs 28b4b2b0-149c-49d2-bacf-74c80e9a51d7`

## Logs:
- **db-cluster-28b4b2b0.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/28b4b2b0-149c-49d2-bacf-74c80e9a51d7/20230518_113349/db-cluster-28b4b2b0.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/28b4b2b0-149c-49d2-bacf-74c80e9a51d7/20230518_113349/db-cluster-28b4b2b0.tar.gz)
- **sct-runner-events-28b4b2b0.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/28b4b2b0-149c-49d2-bacf-74c80e9a51d7/20230518_113349/sct-runner-events-28b4b2b0.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/28b4b2b0-149c-49d2-bacf-74c80e9a51d7/20230518_113349/sct-runner-events-28b4b2b0.tar.gz)
- **sct-28b4b2b0.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/28b4b2b0-149c-49d2-bacf-74c80e9a51d7/20230518_113349/sct-28b4b2b0.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/28b4b2b0-149c-49d2-bacf-74c80e9a51d7/20230518_113349/sct-28b4b2b0.log.tar.gz)

[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-master/job/artifacts-offline-install/job/artifacts-rocky9-nonroot-test/9/)
</details>

**Comment by yaronkaikov** [comment]
@fruch See the comment from @syuu1228, I see in the job history this job never passed. Maybe it's some test issue?

**Comment by yaronkaikov** [comment]
Did we ever manage to make it pass? in staging maybe?

**Comment by fruch** [comment]
> Did we ever manage to make it pass? in staging maybe?

the test passes, the only issue here it's coredump when trying to get scylla version, the error is quite clear.

as I said, one should run the test with keep, and investigate the reason.

test is passing on non-root installs and on installed version from RPM.

**Comment by yaronkaikov** [comment]
@fruch I have connected to the DB-node (https://jenkins.scylladb.com/job/scylla-master/job/artifacts-offline-install/job/artifacts-rocky9-nonroot-test/11/console)
```
[yaronkaikov@london]~$ ssh -i .ssh/scylla-test scylla-test@34.73.145.44
Last login: Tue May 23 11:26:55 2023 from 199.203.229.89
[scylla-test@artifacts-rocky9-jenkins-db-node-ff7c45fd-0-1 ~]$ cd scylladb/
[scylla-test@artifacts-rocky9-jenkins-db-node-ff7c45fd-0-1 scylladb]$ cd bin/
[scylla-test@artifacts-rocky9-jenkins-db-node-ff7c45fd-0-1 bin]$ ./scylla --version
5.3.0~dev-0.20230517.302a89488a28
[scylla-test@artifacts-rocky9-jenkins-db-node-ff7c45fd-0-1 bin]$ 
```

So it's working. i don't understand why the test failed. 

Also not sure if it's related, but at the beginning I couldn't get access to the DB node since the public key was wrong
```
[scylla-test@artifacts-rocky9-jenkins-db-node-ff7c45fd-0-1 bin]$ cat ~/.ssh/authorized_keys 
# Added by Google
ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIIbmiadAJlRDbIHNNK3IYGeuSBM9QeBH7vf1CZ+4LXwF
```

Only after @fgelcer connected via Google console and updated the autorized_keys i manage to connect
Any idea?

**Comment by fruch** [comment]
once scylla is running, then `scylla --version` crashes
```
🧱 INT❯ ssh -i ~/.ssh/scylla-test scylla-test@34.73.145.44
Last login: Tue May 23 11:30:46 2023 from 199.203.229.89
[scylla-test@artifacts-rocky9-jenkins-db-node-ff7c45fd-0-1 ~]$
[scylla-test@artifacts-rocky9-jenkins-db-node-ff7c45fd-0-1 ~]$
[scylla-test@artifacts-rocky9-jenkins-db-node-ff7c45fd-0-1 ~]$ cd scylladb/bin/
[scylla-test@artifacts-rocky9-jenkins-db-node-ff7c45fd-0-1 bin]$ ./scylla --version
5.3.0~dev-0.20230517.302a89488a28
[scylla-test@artifacts-rocky9-jenkins-db-node-ff7c45fd-0-1 bin]$ systemctl --user start scylla-server.service
[scylla-test@artifacts-rocky9-jenkins-db-node-ff7c45fd-0-1 bin]$ ./scylla --version
terminate called after throwing an instance of 'std::runtime_error'
  what():  Could not setup Async I/O: Resource temporarily unavailable. The most common cause is not enough request capacity in /proc/sys/fs/aio-max-nr. Try increasing that number or reducing the amount of logical CPUs available for your application
Aborted (core dumped)
[scylla-test@artifacts-rocky9-jenkins-db-node-ff7c45fd-0-1 bin]$
```

**Comment by fruch** [comment]
> @fruch I have connected to the DB-node (https://jenkins.scylladb.com/job/scylla-master/job/artifacts-offline-install/job/artifacts-rocky9-nonroot-test/11/console)
> 
> ```
> [yaronkaikov@london]~$ ssh -i .ssh/scylla-test scylla-test@34.73.145.44
> Last login: Tue May 23 11:26:55 2023 from 199.203.229.89
> [scylla-test@artifacts-rocky9-jenkins-db-node-ff7c45fd-0-1 ~]$ cd scylladb/
> [scylla-test@artifacts-rocky9-jenkins-db-node-ff7c45fd-0-1 scylladb]$ cd bin/
> [scylla-test@artifacts-rocky9-jenkins-db-node-ff7c45fd-0-1 bin]$ ./scylla --version
> 5.3.0~dev-0.20230517.302a89488a28
> [scylla-test@artifacts-rocky9-jenkins-db-node-ff7c45fd-0-1 bin]$ 
> ```
> 
> So it's working. i don't understand why the test failed.
> 
> Also not sure if it's related, but at the beginning I couldn't get access to the DB node since the public key was wrong
> 
> ```
> [scylla-test@artifacts-rocky9-jenkins-db-node-ff7c45fd-0-1 bin]$ cat ~/.ssh/authorized_keys 
> # Added by Google
> ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIIbmiadAJlRDbIHNNK3IYGeuSBM9QeBH7vf1CZ+4LXwF
> ```
> 
> Only after @fgelcer connected via Google console and updated the autorized_keys i manage to connect Any idea?

your scylla-test key is `ssh-ed25519` and not `ssh-rsa` ?

**Comment by fruch** [comment]
> > Only after @fgelcer connected via Google console and updated the autorized_keys i manage to connect Any idea?
> 
> your scylla-test key is `ssh-ed25519` and not `ssh-rsa` ?

my mistake the needed key is `~/.ssh/scylla_test_id_ed25519` I've forgot about it. that for rocky9 I've stop using rsa key

**Comment by fgelcer** [comment]
> my mistake the needed key is `~/.ssh/scylla_test_id_ed25519` I've forgot about it. that for rocky9 I've stop using rsa key

i tried to connect myself, and couldn't... from the GCP console i could SSH, and added the "scylla test" key to the authorized keys, and then it allowed me in

**Comment by mykaul** [comment]
Perhaps strace to the rescue?

**Comment by syuu1228** [comment]
> [scylla-test@artifacts-rocky9-jenkins-db-node-ff7c45fd-0-1 bin]$ ./scylla --version
> 5.3.0~dev-0.20230517.302a89488a28
> [scylla-test@artifacts-rocky9-jenkins-db-node-ff7c45fd-0-1 bin]$ systemctl --user start scylla-server.service
> [scylla-test@artifacts-rocky9-jenkins-db-node-ff7c45fd-0-1 bin]$ ./scylla --version
> terminate called after throwing an instance of 'std::runtime_error'
>   what():  Could not setup Async I/O: Resource temporarily unavailable. The most common cause is not enough request capacity in /proc/sys/fs/aio-max-nr. Try increasing that number or reducing the amount of logical CPUs available for your application
> Aborted (core dumped)

In this case probably scylla-server.service used all available AIO slots since fs.aio-max-nr size wasn't large enough.
Prevously, when fs.aio-max-nr is small scylla-server just crashed (#9096), so I modified the AIO slot allocator to shrink allocation size to fit to fs.aio-max-nr (https://github.com/scylladb/seastar/commit/2a3b4ac62865644cf2c49840cd3eebf6b92dfd11#).
So it likely no AIO slot available after scylla-server.service started.

We can verify this by checking journal log, if there are the message like "Requested AIO slots too large, please increase request capacity in /proc/sys/fs/aio-max-nr. available:{} requested:{}", then fs.aio-max-nr is not enough and scylla-server.service using all of them.

Possible workaround without changing scylla code are:
 - Increase fs.aio-max-nr
 - Run scylla --version before starting scylla-server.service

Also, we may able to change --version implementation not to run initialization of seastar.

**Comment by yaronkaikov** [comment]
> > [scylla-test@artifacts-rocky9-jenkins-db-node-ff7c45fd-0-1 bin]$ ./scylla --version
> > 5.3.0~dev-0.20230517.302a89488a28
> > [scylla-test@artifacts-rocky9-jenkins-db-node-ff7c45fd-0-1 bin]$ systemctl --user start scylla-server.service
> > [scylla-test@artifacts-rocky9-jenkins-db-node-ff7c45fd-0-1 bin]$ ./scylla --version
> > terminate called after throwing an instance of 'std::runtime_error'
> > what():  Could not setup Async I/O: Resource temporarily unavailable. The most common cause is not enough request capacity in /proc/sys/fs/aio-max-nr. Try increasing that number or reducing the amount of logical CPUs available for your application
> > Aborted (core dumped)
> 
> In this case probably scylla-server.service used all available AIO slots since fs.aio-max-nr size wasn't large enough. Prevously, when fs.aio-max-nr is small scylla-server just crashed (#9096), so I modified the AIO slot allocator to shrink allocation size to fit to fs.aio-max-nr ([scylladb/seastar@2a3b4ac](https://github.com/scylladb/seastar/commit/2a3b4ac62865644cf2c49840cd3eebf6b92dfd11)). So it likely no AIO slot available after scylla-server.service started.
> 
> We can verify this by checking journal log, if there are the message like "Requested AIO slots too large, please increase request capacity in /proc/sys/fs/aio-max-nr. available:{} requested:{}", then fs.aio-max-nr is not enough and scylla-server.service using all of them.
> 
> Possible workaround without changing scylla code are:
> 
> * Increase fs.aio-max-nr
> * Run scylla --version before starting scylla-server.service
> 
> Also, we may able to change --version implementation not to run initialization of seastar.
I guess this make more sense. @mykaul @avikivity Any thoughts ?

**Comment by mykaul** [comment]
> > Also, we may able to change --version implementation not to run initialization of seastar.
> 
> I guess this makes more sense. @mykaul @avikivity Any thoughts?

That's what I thought exactly. Any other flags that are NOT really running Scylla should be initializing much.

**Comment by syuu1228** [comment]
> So I guess something recently added is causing this issue, but it is not in main() function, somewhere else.

Okay I found this is the reason:
https://github.com/scylladb/scylladb/blob/master/test/perf/perf_fast_forward.cc#L1213
`app_template` is allocated on global variable, so Seastar initalized before handling `--version` on `main()`.

**Comment by fruch** [comment]
Happening on 5.3 branch...

**Comment by DoronArazii** [comment]
> Happening on 5.3 branch...

But fixed on master - 5.4
So we need to backport the fix

**Comment by denesb** [comment]
Backported to 5.3.
